File: go/0078M-子集.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-13 20:13:47
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func subsets(nums []int) [][]int {
	ans := make([][]int, 0)
	sub := make([]int, 0)
	var dfs func(index int)
	dfs = func(index int) {
		tmp := make([]int, len(sub))
		copy(tmp, sub)
		ans = append(ans, tmp)
		if index >= len(nums) {
			return
		}
		for i := index; i < len(nums); i++ {
			sub = append(sub, nums[i])
			dfs(i + 1)
			sub = sub[:len(sub)-1]
		}
	}
	dfs(0)
	return ans
}

File: cpp/0078M-子集.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-27 14:37:36
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    vector<vector<int>> ans;
    vector<int> tmp;

    void dfs(vector<int>& nums, int idx) {
        ans.push_back(tmp);
        if (idx >= nums.size()) return;
        for (int i = idx; i < nums.size(); i++) {
            tmp.push_back(nums[i]);
            dfs(nums, i + 1);
            tmp.pop_back();
        }
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        dfs(nums, 0);
        return ans;
    }
};